src/pixie.c

```diff
@@ -18,20 +18,138 @@ void pixie_format(const unsigned char *in, unsigned len, char *outbuf) {
 void pixie_attack(void) {
 	struct pixie *p = &pixie;
 	int dh_small = get_dh_small();
+	int ret_val;
 
 	if(p->do_pixie) {
 		char cmd[4096];
 		snprintf(cmd, sizeof cmd,
 		"pixiewps -e %s -s %s -z %s -a %s -n %s %s %s",
 		p->pke, p->ehash1, p->ehash2, p->authkey, p->enonce,
 		dh_small ? "-S" : "-r" , dh_small ? "" : p->pkr);
-		printf("executing %s\n", cmd);
-		exit(system(cmd));
+		if (p->do_pixie == 2) {
+			ret_val = pixie_wrapper(cmd);
+			if (ret_val != EXIT_SUCCESS) {
+				/* <sys/wait.h> WEXITSTATUS() or div 256 */
+				exit(ret_val);
+			}
+		}
+		else {
+			printf("executing %s\n", cmd);
+			exit(system(cmd));
+		}
 	}
+}
+
+void pixie_free() {
 	PIXIE_FREE(authkey);
 	PIXIE_FREE(pkr);
 	PIXIE_FREE(pke);
 	PIXIE_FREE(enonce);
 	PIXIE_FREE(ehash1);
 	PIXIE_FREE(ehash2);
+	PIXIE_FREE(wrapper);
+	PIXIE_FREE(pin);
+}
+
+/**
+ * Call a wrapper to run pixiewps.
+ * 
+ * The wrapper could be implemented in any language which needs to follow the following conventions:
+ * 1) arguments are passed precisely as for pixiewps
+ * 2) on failure, non-zero exit status is returned
+ * 3) on success, exactly one line is printed to stdout, containing the found pin and a trailing newline character,
+ *    in case of null pin, only a newline character '\n' would be returned and exit status 0 is returned.
+ * 
+ * @params char* cmd The command line to call pixiewps with the arguments
+ * 
+ * @return int
+ */
+int pixie_wrapper(char *cmd) {
+	char *sh = NULL;
+	int ret_val = EXIT_FAILURE;
+
+	sh = (char*) malloc(strlen(pixie.wrapper) + strlen(cmd));
+	if (sh) {
+		/* Call the wrapper with the arguments passed */
+		strcpy(sh, pixie.wrapper);
+		strcat(sh, " ");
+		strcat(sh, cmd+9);
+		printf("executing %s\n", sh);
+		ret_val = pixie_popen(sh);
+		free(sh);
+		sh = NULL;
+	}
+
+	return ret_val;
+}
+
+/**
+ * Executing pixiewps optimized with popen(), set the discovered pin to pixie.pin if Pixiewps attack is successful
+ * 
+ * 1) on failure, non-zero exit status is returned
+ * 2) on success, print the pin and exit status 0 is returned
+ * 
+ * @params char* cmd The command line to call a wrapper file
+ * 
+ * @return int
+ */
+int pixie_popen(char *cmd) {
+	FILE *fp;
+	char *str_pin = NULL;
+	char msg[256];
+	int ret_val = EXIT_FAILURE;
+
+	fp = popen(cmd, "r");
+	if (fp != NULL) {
+		while (fgets(msg, 256, fp) != NULL) {
+			printf("%s", msg);
+			/* Capture pin if the wrapper not following the specifications */
+			if(strstr(msg, "[+] WPS pin: ") != NULL) {
+				str_pin = (char*) malloc(10);
+				if (str_pin) {
+					str_pin[0] = '\0';
+					if (strstr(msg, "empty") == NULL) {
+						strncpy(str_pin, msg+15, 8);
```

Comment:
are you aware what strncpy does ?

Reply:
The key output line of pixiewps when is successful are:

	 [+] WPS pin:  12345670
	 [+] WPS pin:  <empty>

if not found "empty", then the pin is 8 digits.

Reply:
not my question. strncpy has some properties almost nobody is aware of, that makes it unsuitable for 99% of usecases where it is encountered. thus usage of strncpy is a code smell.

from `man 3p strncpy`:
```
     char *strncpy(char *restrict s1, const char *restrict s2, size_t n);
...

     The  stpncpy()  and  strncpy() functions shall copy not more than n bytes
     (bytes that follow a NUL character are not copied) from the array pointed
     to by s2 to the array pointed to by s1.

     If  the  array pointed to by s2 is a string that is shorter than n bytes,
     NUL characters shall be appended to the copy in the array pointed  to  by
     s1, until n bytes in all are written.

...

     If there is no NUL character byte in the  first  n  bytes  of  the  array
     pointed to by s2, the result is not null-terminated.
```

Reply:
I know. So the null character was appended manually.